File: T1.SqlSharp/T1.SqlSharp/Expressions/SqlCaseExpression.cs
using T1.Standard.IO;

namespace T1.SqlSharp.Expressions;

public class SqlCaseClause : ISqlExpression
{
    public SqlType SqlType { get; } = SqlType.CaseClause;
    public TextSpan Span { get; set; } = new();
    public ISqlExpression? Case { get; set; }
    public List<SqlWhenThenClause> WhenThens { get; set; } = [];
    public ISqlExpression? Else { get; set; }

    public string ToSql()
    {
        var sql = new IndentStringBuilder();
        sql.Write("CASE");
        if (Case != null)
        {
            sql.Write(" " + Case.ToSql());
        }
        sql.WriteLine();
        sql.Indent++;
        foreach (var when in WhenThens)
        {
            sql.WriteLine(when.ToSql());
        }
        sql.Indent--;
        if (Else != null)
        {
            sql.WriteLine("ELSE " + Else.ToSql());
        }
        sql.WriteLine("END");
        return sql.ToString();
    }
}

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlCaseExpression.ts
import { SqlType } from './SqlType';
import { TextSpan } from '../StringParser';
import { ISqlExpression, SqlVisitor } from './ISqlExpression';

export class SqlCaseExpression implements ISqlExpression {
    SqlType: SqlType = SqlType.CaseClause;
    Span: TextSpan = new TextSpan();
    Case?: ISqlExpression;
    WhenThens: ISqlExpression[] = [];
    Else?: ISqlExpression;

    Accept(visitor: SqlVisitor): void {
        visitor.Visit_CaseClause(this);
    }

    ToSql(): string {
        let sql = 'CASE ';
        if (this.Case) sql += this.Case.ToSql() + ' ';
        sql += this.WhenThens.map(x => x.ToSql()).join(' ');
        if (this.Else) sql += ' ELSE ' + this.Else.ToSql();
        sql += ' END';
        return sql;
    }
} 